### DISCUSSION TITLE
[Request] New ItemComponent Subtypes Definitions Assistance

### DISCUSSION THREAD

**Discussion by MapleWheels** [FakeFishGames/Barotrauma#10159] [Code questions and modding support]
Good day,

I am aware that this request falls outside the purview of the Dev team and that any issues that arise from what I'm trying to do have no bearing of responsibility on Undertow, Fakefish, and others nor do you have any obligations to help me.

So, using the following mod "CsForBarotrauma", I am attempting to add a new child Type of Reactor in order to implement a new type of power generation (I need to override some of your methods). What I would like help with is understanding your Content Loading Pipeline. I have looked through the source and have a background in .NET so I am aware of how you're using XML (System.Xml) for this. However, I have been told by multiple people that Node Types are "hardcoded" so adding any new ItemComponent Subtypes is highly not advisable. However, I cannot for the life of me see where and how this is. It seems that so long as implement the interfaces and attributes as you have, simply creating the Subtype will make this work (Reflection).

People have recommended using runtime editing, like Harmony, to override methods and just check if it's "my item" but I find that can get messy, especially if multiple people are trying to edit this Type. I think a cleaner approach would be to define my own Subtype and implement the changes I want for my mod there, keeping my changes contained and easy for potential other modders to target if they so wish.

What would be your recommended way to create a new ItemComponent Type to reduce potential conflicts and if you have the time, would you mind explaining/detailing from the code side, your XML content loading pipeline?

Edit: I guess the best way to phrase this question would be "When you guys add a new Subtype to the game for use in XML content (ie. \<PowerContainer\>, \<Powered\>, etc.), what do you guys have to do?" As far as I can see, reflection is doing the work.

Thanks for any assistance you can provide.

Sincerely,

UPDATE: So from what I understand, LuaCsForBarotrauma does not execute any client-side code before you enter a lobby (SubEditor, CampaignLobby, MultiplayerLobby). So I would need to force a reload of all XML data from Content Packages before you join the lobby or you would desync. From what I was told, the method for this was removed in the Modding Refactor Update. Is there any new equivalent for this?

**Comment by Regalis11** [comment]
> Edit: I guess the best way to phrase this question would be "When you guys add a new Subtype to the game for use in XML content (ie. <PowerContainer>, <Powered>, etc.), what do you guys have to do?" As far as I can see, reflection is doing the work.

That's correct: all you need to do is add a new subclass of ItemComponent and the game should be able to find it using reflection.

> UPDATE: So from what I understand, LuaCsForBarotrauma does not execute any client-side code before you enter a lobby (SubEditor, CampaignLobby, MultiplayerLobby). So I would need to force a reload of all XML data from Content Packages before you join the lobby or you would desync. From what I was told, the method for this was removed in the Modding Refactor Update. Is there any new equivalent for this?

I'm not familiar with the inner workings of LuaCsForBarotrauma, so I'm not sure I can answer this. But in the vanilla game, simply enabling a mod will load all the necessary XML content. I'm also not sure which removed method you are referring to.

**Comment by MapleWheels** [comment]
First off, thanks for taking the time to answer this.

The command was one akin to `reloadcorepackage` but for regular packages. This method associated with this command would enable the reloading of XML content.
From talking to the dev of the LuaCsForBarotrauma mod, this is the application load order if client-side code is enabled.

Application Launch:
Step01 > Core Content Loaded
Step02 > Client-side LuaCs Core loaded as part of the above as the primary assembly
Main Menu/Apply Mod Settings:
Step03 > Load Content Packages (includes Lua mods src files)
Lobby Load (Client-Campaign, Multi-SelfHost, SubEditor):
Step04 > Lua/Cs Mods Compiled into their own Assembly and Executed.

So there are a couple of issues in regard to the above. 

The obvious issue is that the ItemComponent subclass will not be added until Step04 after XML loading is completed. I want to know if this will invalidate the node when it's being read while the content is being loaded under Step03, as the subclass doesn't exist yet.

Secondly, as per this function under https://github.com/Regalis11/Barotrauma/blob/4c52fe23a5778d2f2d07d3d1f8cfafca4e57573a/Barotrauma/BarotraumaShared/SharedSource/Utils/ReflectionUtils.cs#L15

```
public static IEnumerable<Type> GetDerivedNonAbstract<T>()
{
	Assembly assembly = typeof(T).Assembly;
	if (!cachedNonAbstractTypes.ContainsKey(assembly))
	{
		cachedNonAbstractTypes[assembly] = (from t in assembly.GetTypes()
			where !t.IsAbstract
			select t).ToImmutableArray();
	}
	return cachedNonAbstractTypes[assembly].Where((Type t) => t.IsSubclassOf(typeof(T)));
}
```

From what I understand from my code dive + discussion with others, the code above will only search the Core assembly (mods are in a different assembly to allow easy loading/unloading) and won't search my assembly for my custom subclass. As a result, I was told to instead hook into the Constructor for `Item` and in post, call `AddComponent()` to add mine manually every time my Item is instanced. Which seems like a workaround that can be avoided.

So the question for this is, is there either; 
A. A method to reload the XML content only for a Regular Package (reloading all will unload the mod Assembly as well) or;
B. To both; 
    1. Have the XML node not invalidated before Step04 and;
    2. Potentially have a changed version of the function above to allow it to search multiple Assemblies (registered in a Dictionary). this would allow Modloaders to have modded Assemblies be searched by Reflection as well.

If the above two things are dealt with, then it will be easy for modders in the future to simply make new subtypes for ItemComponent and expand much more on the game.

Thanks again for your help.

EDIT: If I'm totally off base with XML loading and nodes being lost, then feel free to lecture me :). My ultimate goal is to make it easy for modders to extend the game without having to do the patchwork they are doing now (hooking methods just to implement the logic for their item). This eventually gets really messy when several people are hooking the same method and doing IL patches. if they can do things almost exactly the way the devs do, this simplifies the content creation pipeline and allows modding tools to come in line with in-dev tools.

**Comment by Regalis11** [comment]
Thank you for the clarification! I think I mostly get the issue now.

> A. A method to reload the XML content only for a Regular Package (reloading all will unload the mod Assembly as well) or;

This part I don't quite understand - the method for doing this is enabling the mod (which loads all the content in it). But to me that doesn't seem enough: the cached ItemComponent types wouldn't update when loading XML content. 

I might be completely off here, but to me it sounds like this is more of an issue in LuaCsForBarotrauma than the vanilla game. Since there's no support for adding new ItemComponent classes in the vanilla game, I would expect a mod that adds that support to take this problem into account and e.g. refresh the cached ItemComponents.

> Step04 > Lua/Cs Mods Compiled into their own Assembly and Executed.

This also isn't a part of the vanilla game but something implemented in LuaCsForBarotrauma, so I wonder if there'd be a way to make it do this step before loading the item XML content. But as said, I'm not that familiar with LuaCsForBarotrauma, so there could be something in the vanilla game that makes it difficult or impossible to get it to function in a way that'd allow doing what you're attempting to do.

**Comment by MapleWheels** [comment]
So, the concerns here are split between LuaCsForBarotrauma and Vanilla. On the LCFB side, the ItemComponent reflection lookup will need to be modified to search other assemblies, which the author of the mod said he is willing to do (really simple change). The part that needs to be done on the vanilla side is to either;

A. Allow modders to explicitly reload either XML content or another package at lobby creation time (split the Assembly and XML into two Content Packages so you can reload the latter and not the former).
OR
B. Make it so that invalid nodes found in the XML are not stripped during loading (ItemPrefab) until either the Item is instantiated (lazy resolver) or the lobby is created with a hook to allow code execution before XML nodes are resolved. By this point, any modded assemblies should be loaded and LCFB can ensure that they are checked by ReflectionUtils.

With the above being said, I think that a good way to approach this is to allow modders to define in the filelist XML what needs to be lazy loaded (at Lobby Start) and what does not. I don't think it should be on Barotrauma to determine what needs to be lazy-loaded and what not to. That should be on the developer of the mod, as they know it best. It also frees Barotrauma from being responsible for issues caused by this and having to deal with it, including any weird edge cases needing a very specific load order.

Edit: To answer a question; 
> This also isn't a part of the vanilla game but something implemented in LuaCsForBarotrauma, so I wonder if there'd be a way to make it do this step before loading the item XML content. But as said, I'm not that familiar with LuaCsForBarotrauma, so there could be something in the vanilla game that makes it difficult or impossible to get it to function in a way that'd allow doing what you're attempting to do.

LCFB doesn't add the ability to added new ItemComponents to the game, this is something that's actually supported by native Barotrauma, as it's just Reflection lookup. The only thing LCFB lets you do is run custom Lua and C# scripts as part of your Content Package. 

The problem is that LCFB needs to load c# source files (not compiled assemblies) from the content packages at the same time that all other XML data is being loaded (the same time it is parsed/resolved), so it doesn't have this option. Even if LCFB is a Core Package, the other packages that rely on it are Regular Packages and change with settings.

Once the Lobby is starting, LCFB will take the C# Source files and compile them into assemblies and execute them. The new ItemComponent types would not exist before this point. 

To be clear, the one real issue with the vanilla game LCFB cannot solve is that the game resolves XML content before any modded CSharp Assembly is ever compiled and loaded. All that needs to happen (and this may be a big task tbh) is to allow XML (or other) Content files to be marked for lazy resolving. Lazy resolving would be done at Lobby Start, with two native hooks added that allow code to be executed both right before and right after the lazy loading process. At this point, LCFB and the mods are responsible for handling the rest using those hooks.

**Comment by Regalis11** [comment]
I think I'm either still misunderstanding something here, or the issue is actually somewhere else than what you're suggesting. :/

>A. Allow modders to explicitly reload either XML content or another package at lobby creation time (split the Assembly and XML into two Content Packages so you can reload the latter and not the former).

This is done with the mod load order. If LCFB needs to be loaded before the XML content, that can be enforced with mod load order. But to me that doesn't seem necessary, because:

> B. Make it so that invalid nodes found in the XML are not stripped during loading (ItemPrefab) until either the Item is instantiated (lazy resolver) or the lobby is created with a hook to allow code execution before XML nodes are resolved.

They are not stripped. ItemPrefab doesn't know anything about ItemComponents, nor does it strip any XML elements - the ItemComponents are loaded from the XML when you create an actual item instance. So in a sense, the XML is always resolved lazily. This applies to pretty much all XML content: before the "modding refactor" the order in which different types of content files were loaded was very finicky and differences in the order could break things in various ways, but that has been changed, and now there should be no "hard links" between content types - basically everything is resolved lazily. 

So I doubt that's actually where the underlying problem is: I would suspect there is something wrong with the way LCFB refreshes the cached ItemComponent when loading new C# files.

**Reply by MapleWheels** [reply]
> I think I'm either still misunderstanding something here, or the issue is actually somewhere else than what you're suggesting. :/
> 
> > A. Allow modders to explicitly reload either XML content or another package at lobby creation time (split the Assembly and XML into two Content Packages so you can reload the latter and not the former).
> 
> This is done with the mod load order. If LCFB needs to be loaded before the XML content, that can be enforced with mod load order. But to me that doesn't seem necessary, because:
> 
> > B. Make it so that invalid nodes found in the XML are not stripped during loading (ItemPrefab) until either the Item is instantiated (lazy resolver) or the lobby is created with a hook to allow code execution before XML nodes are resolved.
> 
> They are not stripped. ItemPrefab doesn't know anything about ItemComponents, nor does it strip any XML elements - the ItemComponents are loaded from the XML when you create an actual item instance. So in a sense, the XML is always resolved lazily. This applies to pretty much all XML content: before the "modding refactor" the order in which different types of content files were loaded was very finicky and differences in the order could break things in various ways, but that has been changed, and now there should be no "hard links" between content types - basically everything is resolved lazily.
> 
> So I doubt that's actually where the underlying problem is: I would suspect there is something wrong with the way LCFB refreshes the cached ItemComponent when loading new C# files.

So after spending about 2 hours to re-trace/re-read the code, I was mistaken. In fact, a specific chunk of my notes are just wrong. I think I need to stop doing any code type work after 10pm...

From what I now see, the only thing that needs to be done is to modify `ReflectionUtils.GetDerivedNonAbstract<T>()` to search multiple assemblies from a Dict instead of just the list of types of the native assembly as seen here: `return cachedNonAbstractTypes[assembly].Where(t => t.IsSubclassOf(typeof(T)));` which should be trivial.

Anyways, sorry for the confusion and waste of time.

Edit: I also am wondering if it's worth the performance hit of handling `ReflectionTypeLoadException` in this function. It doesn't seem to occur often so it's probably a non-issue. Just figured I'd mention it.

**Comment by MapleWheels** [comment]
Alright, so I've made the PR to LuaCsForBarotrauma and it's been accepted and everything seems to be working. For reference, here's the PR in case you or anyone else is interested: https://github.com/evilfactory/LuaCsForBarotrauma/pull/117

Feel free to close this issue. Thank you for all of the help you've provided!
